### ISSUE TITLE
'$3' disappears from app-dynamics credentials 

### ISSUE THREAD

**Issue by dflick-pivotal** [cloudfoundry/java-buildpack#381]
`$3` disappears from the `"account-access-key":"abc$3abc"` property on the java command line that starts the app in the container.

1. This is how the user provided service is created: `cf cups app-dynamics -p '{"account-access-key":"abc$3abc", "account-name":"customer1", "application-name":"abc", "host-name":"45.45.45.45", "port":"8888"}'`
1. bind the app with the service
1. restage the app
1. `cf ssh` into the container

`$3` disappears from the `"account-access-key":"abc$3abc"` property on the java command line that starts the app in the container.

```
Griffon:~ dflick$ cf ssh spring-music
vcap@0f9b74bf-4c1b-4679-7028-117af39c3bc4:~$ ps -aef | grep java
vcap 20 0 99 16:45 ? 00:00:36 /home/vcap/app/.java-buildpack/open_jdk_jre/bin/java -Djava.io.tmpdir=/home/vcap/tmp -XX:OnOutOfMemoryError=/home/vcap/app/.java-buildpack/open_jdk_jre/bin/killjava.sh -XX:MaxMetaspaceSize=104857K -Xss349K -Xms681574K -XX:MetaspaceSize=104857K -Xmx681574K -javaagent:/home/vcap/app/.java-buildpack/app_dynamics_agent/javaagent.jar -Dappdynamics.agent.applicationName=abc -Dappdynamics.agent.tierName=spring-music -Dappdynamics.agent.nodeName=spring-music:0 -Dappdynamics.agent.accountAccessKey=abcabc -Dappdynamics.agent.accountName=customer1 -Dappdynamics.controller.hostName=45.45.45.45 -Dappdynamics.controller.port=8888 -cp /home/vcap/app/. org.springframework.boot.loader.JarLauncher
```

This is easy to reproduce. I did with pcf dev and spring-music app.

**Comment by youngm** [comment]
This issue might be of some help: https://github.com/cloudfoundry/java-buildpack/issues/248

**Comment by melo0187** [comment]
Thanks for the detailed reply! 

@nebhale About your suggestion:
On the one hand I fully agree. Using such an approach doesn't seem to be careful.
On the other hand things like the account-access-key for access to a preconfigured AppDynamics controller are beyond my influence.

What I did is to point out how unnecessarily error-prone things can get with special characters in the AppDynamics credentials (and every other service payload for that matter), but that is no guarantee that anyone will pay attention to this suggestion...

About your way of escaping the `$3`:
What we came up with was using the interactive mode and provide the value URL encoded like this:
`abc%243abc`

So in staging_info.yml  (inside the app container) and as argument for java  I see this:
`-Dappdynamics.agent.accountAccessKey=abc%243abc`

Even the AD agent log shows this value:
`Default Agent Account Info Resolver found system property [appdynamics.agent.accountAccessKey] for account access key [abc%243abc]`

However everything is working out just fine this way and we see everything showing up in AppDynamics.

But to be honest I really don't know whether this solution is any better or just as much subject to change and thus just as error-prone as @nebhale 's solution.
To be honest I am not even sure why this is working in the first place...
